src/log/log.c

```diff
@@ -1556,8 +1556,9 @@ __wt_log_allocfile(WT_SESSION_IMPL *session, uint32_t lognum, const char *dest)
     WT_ERR(__log_openfile(session, tmp_id, WT_LOG_OPEN_CREATE_OK, &log_fh));
     WT_ERR(__log_file_header(session, log_fh, NULL, true));
     WT_ERR(__log_prealloc(session, log_fh));
-    WT_ERR(__wt_fsync(session, log_fh, true));
-    WT_ERR(__wt_close(session, &log_fh));
+    if (FLD_ISSET(conn->log_flags, WT_CONN_LOG_ZERO_FILL) || conn->log_extend_len != 0)
+        WT_ERR(__wt_fsync(session, log_fh, true));
```

Comment:
There are several other uses of this function that do not use the file allocated after the call. Therefore, the caller should not assume the file is in any particular state when this call completes.

Reply:
```
int
__wt_log_allocfile(WT_SESSION_IMPL *session, uint32_t lognum, const char *dest)
{
    ......
   **//problem 1**
    WT_ERR(__log_file_header(session, log_fh, NULL, true));----->Here, it will execute __wt_ fsync().
    WT_ERR(__log_prealloc(session, log_fh));
    WT_ERR(__wt_fsync(session, log_fh, true)); -----------> here, it run __wt_ fsync() again.
    ......

     **//problem 2**
    WT_ERR(__wt_close(session, &log_fh));   -------------> first close.
     ......
err:
    WT_ERR(__wt_close(session, &log_fh));   -------------> close again.
   ......
}
```

Sorry, I didn't understand what you meant here? Can you provide me with guidance and explanation.
thanks.
:)

Reply:
The file handle must be closed on some systems in order to do the rename. The other call to `wt_close` closes on error. On success, the 2nd call to `wt_close` is a no-op.
Similarly with `wt_fsync`. The `log_file_header` code must sync out the header as part of its contract. The `log_prealloc` may or may not dirty the file. If it does that must be sync'ed. If the file is not dirty the 2nd call to `wt_fsync` is a no-op.

Having a potential no-op call is easier to reason about than a complex conditional. It makes the guarantees of the function clear whereas a complex conditional means one must think about what cases lead to the condition.

Reply:
got it, thanks.
:)